src/ImageSharp.Web/DependencyInjection/ImageSharpBuilderExtensions.cs

```diff
@@ -123,91 +122,91 @@ public static IImageSharpBuilder SetCacheHash<TCacheHash>(this IImageSharpBuilde
+            where TImplementation : class, IImageProvider
         {
-            builder.Services.TryAddEnumerable(ServiceDescriptor.Singleton<IImageProvider, TProvider>());
+            builder.Services.TryAddEnumerable(ServiceDescriptor.Singleton<IImageProvider, TImplementation>());
+
             return builder;
         }
 
         /// <summary>
         /// Adds the given <see cref="IImageProvider"/> to the provider collection within the service collection.
         /// </summary>
-        /// <typeparam name="TProvider">The type of class implementing <see cref="IImageProvider"/>to add.</typeparam>
+        /// <typeparam name="TImplementation">The type of class implementing <see cref="IImageProvider"/> to add.</typeparam>
         /// <param name="builder">The core builder.</param>
         /// <param name="implementationFactory">The factory method for returning a <see cref="IImageProvider"/>.</param>
         /// <returns>The <see cref="IImageSharpBuilder"/>.</returns>
-        public static IImageSharpBuilder AddProvider<TProvider>(this IImageSharpBuilder builder, Func<IServiceProvider, TProvider> implementationFactory)
-            where TProvider : class, IImageProvider
+        public static IImageSharpBuilder AddProvider<TImplementation>(this IImageSharpBuilder builder, Func<IServiceProvider, TImplementation> implementationFactory)
+            where TImplementation : class, IImageProvider
         {
             builder.Services.TryAddEnumerable(ServiceDescriptor.Singleton<IImageProvider>(implementationFactory));
+
             return builder;
         }
 
         /// <summary>
         /// Inserts the given <see cref="IImageProvider"/> at the give index into to the provider collection within the service collection.
         /// </summary>
-        /// <typeparam name="TProvider">The type of class implementing <see cref="IImageProvider"/>to add.</typeparam>
+        /// <typeparam name="TImplementation">The type of class implementing <see cref="IImageProvider"/> to add.</typeparam>
         /// <param name="builder">The core builder.</param>
         /// <param name="index">The zero-based index at which the provider should be inserted.</param>
         /// <returns>The <see cref="IImageSharpBuilder"/>.</returns>
-        public static IImageSharpBuilder InsertProvider<TProvider>(this IImageSharpBuilder builder, int index)
-            where TProvider : class, IImageProvider
+        public static IImageSharpBuilder InsertProvider<TImplementation>(this IImageSharpBuilder builder, int index)
+            where TImplementation : class, IImageProvider
         {
             var descriptors = builder.Services.Where(x => x.ServiceType == typeof(IImageProvider)).ToList();
-            builder.ClearProviders();
-
-            descriptors.Insert(index, ServiceDescriptor.Singleton<IImageProvider, TProvider>());
+            descriptors.RemoveAll(x => x.GetImplementationType() == typeof(TImplementation));
```

Comment:
Ah right, so you're removing any previous instances of the same type. It's really odd that someone would add multiple instances but I'll go with it.

Reply:
Agreed that it might seem odd, but you don't always own all code and know whether other users have already added a provider. And making the API idempotent is a nice bonus 😉